addon/-private/functional/modifier-manager.ts

```diff
@@ -1,98 +1,101 @@
 import { capabilities } from '@ember/modifier';
 import { gte } from 'ember-compatibility-helpers';
-import type { FunctionalModifierDefinition } from './modifier';
+import type { FunctionalModifierDefinition, Teardown } from './modifier';
 import type { ArgsFor, ElementFor } from '../signature';
 import { consumeArgs, Factory, isFactory } from '../compat';
-import { assert } from '@ember/debug';
 
-type ElementMap = {
-  get<S>(key: FunctionalModifierDefinition<S>): ElementFor<S> | null;
-  set<S>(key: FunctionalModifierDefinition<S>, value: ElementFor<S>): void;
-};
-
-type TeardownMap = {
-  get<S>(key: FunctionalModifierDefinition<S>): unknown;
-  set<S>(key: FunctionalModifierDefinition<S>, value: unknown): void;
-};
-
-const MODIFIER_ELEMENTS = new WeakMap() as ElementMap;
-const MODIFIER_TEARDOWNS = new WeakMap() as TeardownMap;
-
-function teardown<S>(modifier: FunctionalModifierDefinition<S>): void {
-  const teardown = MODIFIER_TEARDOWNS.get(modifier);
+interface State<S> {
+  instance: FunctionalModifierDefinition<S>;
+}
 
-  if (teardown && typeof teardown === 'function') {
-    teardown();
-  }
+interface CreatedState<S> extends State<S> {
+  element: null;
 }
 
-function setup<S>(
-  modifier: FunctionalModifierDefinition<S>,
-  element: ElementFor<S>,
-  args: ArgsFor<S>
-): void {
-  const { positional, named } = args;
-  const teardown = modifier(element, positional, named);
+interface InstalledState<S> extends State<S> {
+  element: ElementFor<S>;
+  teardown?: Teardown;
+}
 
-  MODIFIER_TEARDOWNS.set(modifier, teardown);
+// Wraps the unsafe (b/c it mutates, rather than creating new state) code that
+// TS does not yet understand.
+function installElement<S>(
+  state: CreatedState<S>,
+  element: ElementFor<S>
+): InstalledState<S> {
+  // SAFETY: this cast represents how we are actually handling the state machine
+  // transition: from this point forward in the lifecycle of the modifier, it
+  // always behaves as `InstalledState<S>`. It is safe because, and *only*
+  // because, we immediately initialize `element`. (We cannot create a new state
+  // from the old one because the modifier manager API expects mutation of a
+  // single state bucket rather than updating it at hook calls.)
+  const installedState = state as State<S> as InstalledState<S>;
+  installedState.element = element;
+  return installedState;
 }
 
-class FunctionalModifierManager<S> {
+export default class FunctionalModifierManager<S> {
   capabilities = capabilities(gte('3.22.0') ? '3.22' : '3.13');
 
+  options: { eager: boolean };
+
+  constructor(options?: { eager: boolean }) {
+    this.options = {
+      eager: options?.eager ?? true,
+    };
+  }
+
   createModifier(
     factoryOrClass:
       | Factory<FunctionalModifierDefinition<S>>
       | FunctionalModifierDefinition<S>
-  ): FunctionalModifierDefinition<S> {
-    const Modifier = isFactory(factoryOrClass)
+  ): CreatedState<S> {
+    const instance = isFactory(factoryOrClass)
```

Comment:
minor note here: this ternary should be removed in 4.0 (when we drop < ember-source@3.22 support)

Reply:
Yep, noted the same.